service/storage_proxy.cc

```diff
@@ -4601,6 +4636,16 @@ std::vector<gms::inet_address> storage_proxy::get_live_endpoints(keyspace& ks, c
     return eps;
 }
 
+std::vector<gms::inet_address> storage_proxy::get_live_endpoints_for_read(keyspace& ks, const dht::token& token, clock_type::time_point timeout) {
+    auto& rs = ks.get_replication_strategy();
+    std::vector<gms::inet_address> eps = rs.get_natural_endpoints_without_node_being_replaced(token);
+    auto itend = boost::range::remove_if(eps, [this, timeout, &gossiper = gms::get_local_gossiper()] (const gms::inet_address& ep) {
+        return !gossiper.is_alive(ep) || !is_likely_to_respond_in_time(ep, timeout);
```

Comment:
What if because of all this removal of candidates, we aren't left with enough candidates to fulfill CL? Maybe in this case we have to try those unlikely-to-respond nodes anyway? Or do we intentially want to return an error in this case - as a way to reduce the overload?

Reply:
Right, that's a very valid point. I suppose it may also depend what the client might want (which we could deduce from extended service level information, coming soon). Right now the code just prefers to drop the request fast even if it had a slim chance of succeeding, but it makes more sense to give some leeway if we're about to disqualify a node and it's crucial for CL to be fulfilled. That would render cl==ALL queries immune to this failure detector, but perhaps users sending cl==ALL requests are likely to want the query to succeed no matter what - otherwise, why set such a high consistency level.

I'll rethink the code to add a bypass for this disqualification mechanism if we're about to not fulfill the consistency level.

Reply:
I think we should give each node a score (maybe true/false) and sort by score. This way we don't lose CL.

Reply:
(but still prefer more-live replicas)